include/neml2/models/crystallography/CrystalGeometry.h

```diff
@@ -0,0 +1,168 @@
+class CrystalGeometry : public StaticModel
+{
+public:
+  /// Setup from parameter set
+  CrystalGeometry(const OptionSet & options);
+
+  /// Alternate constructor not relying on options
+  CrystalGeometry(const OptionSet & options,
+                  const CrystalClass & cclass,
+                  const Vec & lattice_vectors);
+
+  /// Input options
+  static OptionSet expected_options();
+
+  /// accessor for the first lattice vector
+  Vec a1() const;
+  /// accessor for the second lattice vector
+  Vec a2() const;
+  /// accessor for the third lattice vector
+  Vec a3() const;
+
+  /// accessor for the first reciprocal lattice vector
+  Vec b1() const;
+  /// accessor for the second reciprocal lattice vector
+  Vec b2() const;
+  /// accessor for the third reciprocal lattice vector
+  Vec b3() const;
+
+  /// Total number of slip systems
+  TorchSize nslip() const;
+  /// Number of slip groups
+  TorchSize nslip_groups() const;
+  /// Number of slip systems in a given group
+  TorchSize nslip_in_group(TorchSize i) const;
+
+  /// Accessor for the slip directions
+  const Vec & cartesian_slip_directions() const { return _cartesian_slip_directions; };
+  /// Accessor for the slip planes
+  const Vec & cartesian_slip_planes() const { return _cartesian_slip_planes; };
+  /// Accessor for the burgers vector
+  const Scalar & burgers() const { return _burgers; };
+
+  /// Accessor for the full Schmid tensors
+  const R2 & A() const { return _A; };
+  /// Accessor for the symmetric Schmid tensors
+  const SR2 & M() const { return _M; };
+  /// Accessor for the skew-symmetric Schmid tensors
+  const WR2 & W() const { return _W; };
+
+  /// Accessor for the crystal class
+  const CrystalClass & crystal_class() const { return _class; };
+
+  /// Slice a BatchTensor to provide only the batch associated with a slip system
+  // The slice happens along the last batch axis
+  template <
+      class Derived,
+      typename = typename std::enable_if_t<std::is_base_of_v<BatchTensorBase<Derived>, Derived>>>
+  Derived slip_slice(const Derived & tensor, TorchSize grp) const;
+
+private:
+  /// Delegated constructor to setup schmid tensors and slice indices at once
+  CrystalGeometry(const OptionSet & options,
+                  const CrystalClass & cclass,
+                  const Vec & lattice_vectors,
+                  std::tuple<Vec, Vec, Scalar, std::vector<TorchSize>> slip_data);
+
+  /// Helper to setup reciprocal lattice
+  static Vec make_reciprocal_lattice(const Vec & lattice_vectors);
+
+  /// Helper to setup slip systems
+  static std::tuple<Vec, Vec, Scalar, std::vector<TorchSize>>
+  setup_schmid_tensors(const Vec & A,
+                       const CrystalClass & cls,
+                       const MillerIndex & slip_directions,
+                       const MillerIndex & slip_planes);
+
+  /// Helper to return Cartesian vector given miller indices
+  static Vec miller_to_cartesian(const Vec & A, const MillerIndex & d);
```

Comment:
Can these be free functions under the `crystallography` namespace?

Reply:
Oh they are static functions so guess this is fine.

Reply:
This whole annoyance is just to let me initialize everything as `const`.  I don't know if it was really worth it, but here we are.